File: tools/ping/tcpingoo.py
import sys
import socket
import time
import argparse
import signal
import os
from PyQt5.QtWidgets import (
    QApplication,
    QMainWindow,
    QWidget,
    QVBoxLayout,
    QHBoxLayout,
    QLabel,
    QLineEdit,
    QPushButton,
    QFormLayout,
    QRadioButton,
    QSpinBox,
    QTextEdit,
)


# 添加此全局标志变量以跟踪是否使用了 Ctrl+C
ctrl_c_used = False

# 检测系统是否有图形界面
def has_gui():
    try:
        from PyQt5.QtWidgets import QApplication
        return True
    except ImportError:
        return False


# 自定义DNS解析异常类，用于抛出更具体的错误信息
def custom_gaierror(msg):
    class CustomGaiError(socket.gaierror):
        def __init__(self, message):
            super().__init__(-1, message)

    raise CustomGaiError(msg)


# SIGINT信号处理程序 (Ctrl+C)
def signal_handler(sig, frame):
    global ctrl_c_used
    ctrl_c_used = True


# 解析目标主机的IP地址，支持IPv4和IPv6
def resolve_ip(hostname, force_ipv4=False):
    try:
        # 默认使用IPv4进行DNS查询
        family = socket.AF_INET if force_ipv4 else socket.AF_INET6

        # 使用系统默认的DNS服务器进行查询
        addr_info = socket.getaddrinfo(hostname, None, family)
        ip = addr_info[0][4][0]  # 获取IP地址

        return ip

    except socket.gaierror:
        raise ValueError(f"TCPing 请求找不到主机 {hostname}。请检查该名称，然后重试.")


# 执行TCP Ping操作，测量响应时间和丢包率
def tcping(
    domain,
    port,
    request_nums,
    force_ipv4,
    force_ipv6,
    timeout=1000,
    continuous_ping=False,
    ttl=64,
):
    try:
        ip = None

        # 根据参数设置DNS解析方式
        if force_ipv4:
            # 如果使用 -4 参数，只使用IPv4进行DNS查询
            ip = resolve_ip(domain, force_ipv4=True)
        elif force_ipv6:
            # 如果使用 -6 参数，只使用IPv6进行DNS查询
            ip = resolve_ip(domain, force_ipv4=False)
        else:
            # 否则，根据系统的网络配置来选择DNS查询方式
            try:
                # 尝试使用IPv4进行DNS查询
                ip = resolve_ip(domain, force_ipv4=True)
            except ValueError:
                # 如果IPv4查询失败，则使用IPv6进行DNS查询
                ip = resolve_ip(domain, force_ipv4=False)

        response_text = f"\n正在 TCPing {domain}:{port} [{ip}:{port}] 具有 32 字节的数据:\n"
        request_num = 1
        response_times = []
        received_count = 0
        lost_count = 0

        try:
            while not continuous_ping and request_num <= request_nums:
                if ctrl_c_used:  # 检查是否使用了 Ctrl+C
                    break

                start_time = time.time()
                try:
                    with socket.create_connection(
                        (ip, port), timeout=timeout / 1000
                    ) as conn:
                        # 在发送 ping 请求之前设置 TTL 值
                        conn.setsockopt(socket.IPPROTO_IP, socket.IP_TTL, ttl)

                        end_time = time.time()
                        response_time = (end_time - start_time) * 1000  # 转换为毫秒
                        response_times.append(response_time)
                        response_text += (
                            f"来自 {ip}:{port} 的回复: 字节=32 时间={response_time:.0f}ms TTL={ttl}\n"
                        )
                        received_count += 1
                        request_num += 1
                        time.sleep(1)
                except socket.timeout:
                    response_text += "请求超时。\n"
                    lost_count += 1
                    request_num += 1
                    time.sleep(1)
                except (OSError, ConnectionRefusedError) as e:
                    if isinstance(e, OSError) and e.errno == 10049:
                        response_text += "请求超时。\n"
                        lost_count += 1
                        request_num += 1
                        time.sleep(1)
                    else:
                        response_text += f"无法连接到 {ip}:{port}。\n"
                        lost_count += 1
                        request_num += 1
                        time.sleep(1)

        except KeyboardInterrupt:
            pass

        # 计算发送的总数据包数（request_num - 1）
        total_packets_sent = request_num - 1

        # 计算丢包率，考虑没有接收到数据包的情况
        if total_packets_sent > 0:
            packet_loss_rate = (lost_count / total_packets_sent) * 100
        else:
            packet_loss_rate = 100.0

        avg_delay = sum(response_times) / received_count if received_count > 0 else 0.0
        min_delay = min(response_times) if received_count > 0 else 0.0
        max_delay = max(response_times) if received_count > 0 else 0.0

        response_text += (
            f"\n{ip}:{port} 的 TCPing 统计信息:\n"
            f"    数据包: 已发送 = {total_packets_sent}, 已接收 = {received_count}，丢失 = {lost_count} ({packet_loss_rate:.1f}% 丢失)\n"
        )

        if received_count > 0:
            response_text += "往返行程的估计时间(以毫秒为单位):\n"
            response_text += (
                f"    最短 = {min_delay:.0f}ms，最长 = {max_delay:.0f}ms，平均 = {avg_delay:.0f}ms\n"
            )
        else:
            response_text += "所有请求均超时，无法计算往返行程时间.\n"

        if ctrl_c_used:  # 如果使用 Ctrl+C，则仅打印“Control-C”
            response_text += "Control-C\n"

        return response_text

    except ValueError as e:
        return str(e)


# 主函数
def main():
    if not has_gui() or len(sys.argv) > 1:
        # If there is no GUI support or command-line arguments are provided, use CLI mode.
        script_name = os.path.basename(sys.argv[0])  # 获取脚本或可执行文件名称

        parser = argparse.ArgumentParser(
            description=f"{script_name} - 使用 TCP 协议检查目标主机端口的可达性。",
            formatter_class=argparse.RawDescriptionHelpFormatter,
            epilog="示例:\n"
            f"{script_name} example.com 80\n"
            f"{script_name} example.com 80 -4\n"
            f"{script_name} example.com 80 -6\n"
            f"{script_name} example.com 80 -d 1.1.1.1\n"
            f"{script_name} example.com 80 -h\n"
            f"{script_name} example.com 80 -i 128\n"
            f"{script_name} example.com 80 -n 4\n"
            f"{script_name} example.com 80 -w 1000",
            usage="%(prog)s domain port [-4] [-6] [-d DNS_server] [-h] [-i TTL] [-n count] [-t] [-w timeout]",
            add_help=False,
        )

        parser.add_argument("domain", help="要 TCPing 的目标主机名。")
        parser.add_argument("port", type=int, help="目标主机的端口号。")
        parser.add_argument("-4", dest="force_ipv4", action="store_true", help="强制使用 IPv4。")
        parser.add_argument("-6", dest="force_ipv6", action="store_true", help="强制使用 IPv6。")
        parser.add_argument(
            "-d",
            dest="dns_server",
            metavar="DNS_server",
            default=None,
            help="自定义 DNS 服务器地址。",
        )
        parser.add_argument("-h", action="help", help="显示帮助信息并退出。")
        parser.add_argument(
            "-i", dest="ttl", metavar="TTL", type=int, default=128, help="生存时间。"
        )
        parser.add_argument(
            "-n",
            dest="request_nums",
            metavar="count",
            type=int,
            default=4,
            help="要发送的回显请求数。",
        )
        parser.add_argument(
            "-t",
            dest="continuous_ping",
            action="store_true",
            help="Ping 指定的主机，直到停止。\n若要查看统计信息并继续操作，请键入 Ctrl+Break； \n若要停止，请键入 Ctrl+C。",
        )
        parser.add_argument(
            "-w",
            dest="timeout",
            metavar="timeout",
            type=int,
            default=1000,
            help="等待每次回复的超时时间(毫秒)。",
        )

        args = parser.parse_args()

        try:
            if args.request_nums < 1:
                args.request_nums = 4

            if args.continuous_ping and args.request_nums > 0:
                print("[警告]：同时使用 -t 和 -n 参数时，-t 参数将被忽略，仅执行指定次数的 TCP Ping 操作。")

            result = tcping(
                args.domain,
                args.port,
                args.request_nums,
                args.force_ipv4,
                args.force_ipv6,
                args.timeout,
                args.continuous_ping and args.request_nums == 0,  # 仅在不设置 -n 参数时生效
                args.ttl,
            )

            print(result)

        except ValueError as e:
            print(e)

    else:
        # Use GUI mode if no command-line arguments are provided and there is GUI support.
        app = QApplication(sys.argv)
        window = MainWindow()
        window.show()
        sys.exit(app.exec_())


# 创建主窗口类
class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("TCPingoo - TCP Ping 工具")
        self.setGeometry(100, 100, 600, 400)
        self.central_widget = QWidget()
        self.setCentralWidget(self.central_widget)

        self.create_widgets()
        self.layout_widgets()

    def create_widgets(self):
        self.host_label = QLabel("主机名:")
        self.host_input = QLineEdit()

        self.port_label = QLabel("端口号:")
        self.port_input = QSpinBox()
        self.port_input.setMinimum(1)
        self.port_input.setMaximum(65535)

        self.ipv4_radio = QRadioButton("强制使用 IPv4")
        self.ipv6_radio = QRadioButton("强制使用 IPv6")

        self.request_nums_label = QLabel("回显请求数:")
        self.request_nums_input = QSpinBox()
        self.request_nums_input.setMinimum(1)

        self.timeout_label = QLabel("超时时间 (毫秒):")
        self.timeout_input = QSpinBox()
        self.timeout_input.setMinimum(1)

        self.ping_button = QPushButton("执行 TCP Ping")
        self.ping_button.clicked.connect(self.execute_tcp_ping)

        self.result_text = QTextEdit()
        self.result_text.setReadOnly(True)

    def layout_widgets(self):
        form_layout = QFormLayout()
        form_layout.addRow(self.host_label, self.host_input)
        form_layout.addRow(self.port_label, self.port_input)
        form_layout.addRow(self.ipv4_radio, self.ipv6_radio)
        form_layout.addRow(self.request_nums_label, self.request_nums_input)
        form_layout.addRow(self.timeout_label, self.timeout_input)
        form_layout.addWidget(self.ping_button)

        layout = QVBoxLayout()
        layout.addLayout(form_layout)
        layout.addWidget(self.result_text)

        self.central_widget.setLayout(layout)

    def execute_tcp_ping(self):
        domain = self.host_input.text()
        port = self.port_input.value()
        force_ipv4 = self.ipv4_radio.isChecked()
        force_ipv6 = self.ipv6_radio.isChecked()
        request_nums = self.request_nums_input.value()
        timeout = self.timeout_input.value()

        result = tcping(
            domain,
            port,
            request_nums,
            force_ipv4,
            force_ipv6,
            timeout=timeout,
        )

        self.result_text.setPlainText(result)


if __name__ == "__main__":
    # 设置 SIGINT 的信号处理程序 (Ctrl+C)
    signal.signal(signal.SIGINT, signal_handler)
    main()

File: tools/ping/tcpingoo.go
package main

import (
	"fmt"
	"net"
	"os"
	"time"
)

// ctrlCUsed 是一个全局标志变量，用于跟踪是否使用了Ctrl+C
var ctrlCUsed = false

// signalHandler 处理SIGINT信号（Ctrl+C）
func signalHandler(sig os.Signal) {
	ctrlCUsed = true
}

// resolveIP 解析目标主机的IP地址，支持IPv4和IPv6，并添加自定义DNS服务器支持
func resolveIP(hostname string, forceIPv4, forceIPv6 bool, dnsServer string) (string, error) {
	family := ""
	if forceIPv4 {
		family = "ip4"
	} else if forceIPv6 {
		family = "ip6"
	}

	addrInfo, err := net.LookupIP(hostname)
	if err != nil {
		return "", err
	}

	ip := ""
	for _, addr := range addrInfo {
		if family == "" || addr.String() == family {
			ip = addr.String()
			break
		}
	}

	if ip == "" {
		return "", fmt.Errorf("TCPing 请求找不到主机 %s。请检查该名称，然后重试.", hostname)
	}

	return ip, nil
}

// tcping 执行TCP Ping操作，测量响应时间和丢包率
func tcping(domain string, port int, requestNums int, forceIPv4, forceIPv6 bool, timeout time.Duration, continuousPing bool, ttl int) {
	ip, err := resolveIP(domain, forceIPv4, forceIPv6, "")
	if err != nil {
		fmt.Println(err)
		return
	}

	fmt.Printf("\n正在 TCPing %s:%d [%s:%d] 具有 32 字节的数据:\n", domain, port, ip, port)
	requestNum := 1
	responseTimes := []float64{}
	receivedCount := 0
	lostCount := 0

	for continuousPing || requestNum <= requestNums {
		if ctrlCUsed { // 检查是否使用了 Ctrl+C
			break
		}

		startTime := time.Now()
		conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%d", ip, port), timeout)
		if err == nil {
			defer conn.Close()

			err = conn.SetDeadline(time.Now().Add(timeout))
			if err != nil {
				fmt.Println("设置连接超时时间时出错:", err)
				break
			}

			endTime := time.Now()
			responseTime := endTime.Sub(startTime).Seconds() * 1000 // 转换为毫秒
			responseTimes = append(responseTimes, responseTime)
			fmt.Printf("来自 %s:%d 的回复: 字节=32 时间=%.0fms TTL=%d\n", ip, port, responseTime, ttl)
			receivedCount++
		} else {
			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
				fmt.Println("请求超时。")
			} else {
				fmt.Printf("无法连接到 %s:%d。\n", ip, port)
			}
			lostCount++
		}

		requestNum++
		time.Sleep(1 * time.Second)
	}

	// 计算发送的总数据包数（requestNum - 1）
	totalPacketsSent := requestNum - 1

	// 计算丢包率，考虑没有接收到数据包的情况
	packetLossRate := float64(lostCount) / float64(totalPacketsSent) * 100.0

	// 计算平均、最小和最大延迟
	avgDelay := sum(responseTimes) / float64(receivedCount)
	minDelay := min(responseTimes)
	maxDelay := max(responseTimes)

	fmt.Printf("\n%s:%d 的 TCPing 统计信息:\n", ip, port)
	fmt.Printf("    数据包: 已发送 = %d, 已接收 = %d，丢失 = %d (%.1f%% 丢失)\n", totalPacketsSent, receivedCount, lostCount, packetLossRate)
	if receivedCount > 0 {
		fmt.Printf("往返行程的估计时间(以毫秒为单位):\n")
		fmt.Printf("    最短 = %.0fms，最长 = %.0fms，平均 = %.0fms\n", minDelay, maxDelay, avgDelay)
	} else {
		fmt.Println("所有请求均超时，无法计算往返行程时间.")
	}

	if ctrlCUsed { // 如果使用 Ctrl+C，则仅打印“Control-C”
		fmt.Println("Control-C")
	}
}

func sum(values []float64) float64 {
	total := 0.0
	for _, v := range values {
		total += v
	}
	return total
}

func min(values []float64) float64 {
	if len(values) == 0 {
		return 0
	}
	minValue := values[0]
	for _, v := range values {
		if v < minValue {
			minValue = v
		}
	}
	return minValue
}

func max(values []float64) float64 {
	if len(values) == 0 {
		return 0
	}
	maxValue := values[0]
	for _, v := range values {
		if v > maxValue {
			maxValue = v
		}
	}
	return maxValue
}

func main() {
	if len(os.Args) < 3 {
		fmt.Println("用法: tcping <目标主机> <端口> [-4] [-6] [-n 数量] [-t] [-w 超时时间]")
		os.Exit(1)
	}

	domain := os.Args[1]
	port := atoi(os.Args[2])

	requestNums := 4
	timeout := 1000
	continuousPing := false
	forceIPv4 := false
	forceIPv6 := false

	for i := 3; i < len(os.Args); i++ {
		arg := os.Args[i]
		switch arg {
		case "-4":
			forceIPv4 = true
		case "-6":
			forceIPv6 = true
		case "-n":
			if i+1 < len(os.Args) {
				requestNums = atoi(os.Args[i+1])
				i++
			}
		case "-t":
			continuousPing = true
		case "-w":
			if i+1 < len(os.Args) {
				timeout = atoi(os.Args[i+1])
				i++
			}
		default:
			fmt.Println("无效的参数:", arg)
			os.Exit(1)
		}
	}

	tcping(domain, port, requestNums, forceIPv4, forceIPv6, time.Duration(timeout)*time.Millisecond, continuousPing, 128)
}

func atoi(s string) int {
	n := 0
	for _, c := range s {
		n = n*10 + int(c-'0')
	}
	return n
}
